### PR TITLE
Movement relative to neighbours

### LINKED ISSUE #65: Movement (speed and directionality) relative to neighbours
# UserStory

From a biological perspective, the `Movement (speed and directionality) relative to neighbours` of a cell during its life cycle may be a good descriptor for the cell type (e.g. epidermal vs. mesodermal vs. macrophage). Thus, it would be useful to have this information as a feature that may be computed from the cell lineage tree.

# Acceptance criteria
1. Available as Feature within the Feature Computer
1. Computes the Movement Speed and directionality relative to neighbours

# Tasks
- [x] Implement Feature
- [x] Implement FeatureComputer
- [x] Potential approach for relative speed (Spot level):
  - [x] Compute on spot level: n (e.g. 5) nearest neighbors (cf. nearest neighbor statistics feature).
  - [x] Compute displacement length on spot level (equals the speed, since the time frames are assumed constant)
  - [x] Compute average displacement of neighbors.
  - [x] Compute the difference of these two.
- [x] Compute the average of this over all spots of a BranchSpot to get the equivalent measure on BranchSpot level
- [x] Potential approach for relative direction (Spot level):
  - [x] Compute direction vectors (i.e. spot vs. previous spot) of all neighbours.
  - [x] Compute the sum of these and divide by number of neighbours
  - [x] Compute the difference between the spot vector and this sum and normalize the result.
- [x] Compute the average of this over all spots of a BranchSpot

- [x] Write Unit test
- [x] Write Documentation

# To be discussed

* Is anisotropy a problem for the approach? Can it be handled by setting and using scales of the axes?

# Notes

grafik

### PR DESCRIPTION
**Pull request by stefanhahmann**
Resolves #65

### CODE REVIEW

**Review by maarzt — COMMENTED** [review]
I think the code computes the correct results. But I would suggest some refactoring of the `relativeMovement` method and to change the unit test a bit.

File: src/main/java/org/mastodon/mamut/feature/spot/SpotFeatureUtils.java [hunk 1]
```diff
@@ -0,0 +1,141 @@
...
+		search.search( spot );
+		final RefList< Spot > neighbors = RefCollections.createRefList( model.getGraph().vertices() );
+		while ( search.hasNext() && neighbors.size() < n )
+		{
+			Spot neighbor = search.next();
+			if ( neighbor.equals( spot ) || neighbor.incomingEdges().isEmpty() )
+				continue;
+			neighbors.add( neighbor );
+		}
+		return neighbors;
```
[the remark this answers is not in the document]
Comment (stefanhahmann) [reply src/main/java/org/mastodon/mamut/feature/spot/SpotFeatureUtils.java:139]: I decided to change the name to `getNNearestNeightbors` (the `N` was indeed missing) and to include an optional filter that can be applied to the neighbor candidates of the spot.

My feeling is that the re-use potential of a method that directly returns a list of neighbors of a spot is bit higher than returning a search iterator.

File: src/main/java/org/mastodon/mamut/feature/relativemovement/RelativeMovementPlugin.java [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+	public void setAppPluginModel( final ProjectModel projectModel )
+	{
+		this.projectModel = projectModel;
+	}
+
+	@Override
+	public List< ViewMenuBuilder.MenuItem > getMenuItems()
+	{
+		return Collections.singletonList(
+				menu( "Plugins", menu( "Compute", item( DESCRIPTION ) ) ) );
```
[the remark this answers is not in the document]
Comment (stefanhahmann) [reply src/main/java/org/mastodon/mamut/feature/relativemovement/RelativeMovementPlugin.java:79]: Good idea! Changed it.

File: src/main/java/org/mastodon/mamut/feature/relativemovement/RelativeMovementPlugin.java [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+import static org.mastodon.app.ui.ViewMenuBuilder.item;
+import static org.mastodon.app.ui.ViewMenuBuilder.menu;
+
+@SuppressWarnings( "unused" )
+@Plugin( type = MamutPlugin.class )
+public class RelativeMovementPlugin implements MamutPlugin
+{
+	private static final String DESCRIPTION = "Movement of spots relative to nearest neighbors";
+
+	private static final String[] IMAGE_J_KEYS = { "not mapped" };
```
[the remark this answers is not in the document]
Comment (stefanhahmann) [reply src/main/java/org/mastodon/mamut/feature/relativemovement/RelativeMovementPlugin.java:53]: Good idea! Done.

File: src/test/java/org/mastodon/mamut/feature/branch/exampleGraph/ExampleGraph3.java [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+
+/**
+ * Represents a {@link AbstractExampleGraph} with the following {@link ModelGraph} and {@link ModelBranchGraph}:
+ *
+ * <h1>Model-Graph (i.e. Graph of Spots)</h1>
+ * <pre>
+ * Spot( 0, X=1, Y=1, tp=0 )		Spot( 3, X=0, Y=1, tp=0 )		Spot( 0, X=2, Y=1, tp=0 )
+ *              │								 │								 │
+ * Spot( 1, X=1, Y=2, tp=1 )		Spot( 4, X=0, Y=0, tp=1 )	    Spot( 1, X=2, Y=0, tp=1 )
+ *              │								 │								 │
```
[the remark this answers is not in the document]
Comment (stefanhahmann) [reply src/test/java/org/mastodon/mamut/feature/branch/exampleGraph/ExampleGraph3.java]: Thanks for spotting this. Changed it.

File: doc/deep_lineage/features/spotRelativeMovementEquation.txt [hunk 5]
```diff
@@ -0,0 +1,18 @@
...
+\\
+\vec{d_s}=\vec{pos_s} - \vec{pos_{predecessor(s)}}
+\\
+\vec{d_{s,i}}:=\text{direction of movement of i}^{th}\text{ nearest neighbor of s}
+\\
+\vec{r_s}:=\text{relative movement of spot s}
+\\
+\vec{r_s}=\vec{d_s}-\frac{1}{n} \sum_{i=1}^{n}\vec{d_{s,i}}
+\\
+v_s := \text{spot movement distance}
```
[the remark this answers is not in the document]
Comment (stefanhahmann) [reply doc/deep_lineage/features/spotRelativeMovementEquation.txt:16]: Thanks for spotting this! I changed it to `relative movement distance of spot s`

File: doc/deep_lineage/features/branchRelativeMovementEquation.txt [hunk 6]
```diff
@@ -0,0 +1,18 @@
...
+\\
+\vec{r_b} := \sum_{s \in b} \vec{r_s}
+\\
+\vec{n_b} := \text{normalized movement direction of the branch}
+\\
+\vec{n_b} = \frac{\vec{r_b}}{\left | \vec{r_b} \right |}
+\\
+v_b := \text{relative movement distance of branch}
+\\
+v_b = \frac{1}{\left | b \right |} \sum_{s \in b}\vec{v_{s}} \text{  (relative movement distance of branch)}
```
[the remark this answers is not in the document]
Comment (stefanhahmann) [reply doc/deep_lineage/features/branchRelativeMovementEquation.txt:18]: Thanks for spotting. I fixed this.

File: src/main/java/org/mastodon/mamut/feature/spot/SpotFeatureUtils.java [hunk 7]
```diff
@@ -0,0 +1,141 @@
...
+	public static double[] relativeMovement( final Spot spot, final int n, final Model model,
+			final Function< Spot, double[] > movementProvider )
+	{
+		if ( spot == null )
+			throw new IllegalArgumentException( "Spot is null." );
+		if ( n < 1 )
+			throw new IllegalArgumentException( "Number of neighbors must be at least 1." );
+		List< Spot > neighbors = getNearestNeighbors( model, spot, n );
+		if ( neighbors.isEmpty() )
+			return new double[ 0 ];
```
[the remark this answers is not in the document]
Comment (stefanhahmann) [reply src/main/java/org/mastodon/mamut/feature/spot/SpotFeatureUtils.java:97]: I agree. I changed it as suggested.
Btw: SonarLint does not like returning null, but can be told to be quiet in this case by @SuppressWarnings( "squid:S1168" )

File: src/main/java/org/mastodon/mamut/feature/spot/SpotFeatureUtils.java [hunk 8]
```diff
@@ -0,0 +1,141 @@
...
+	 *     <li>If the spot has no predecessor, an empty array is returned.</li>
+	 * </ul>
+	 * @param spot the spot. If null, an IllegalArgumentException is thrown.
+	 * @param n the number of neighbors to consider. Must be at least 1. If less than 1, an IllegalArgumentException is thrown.
+	 * @param model the model
+	 * @param movementProvider the function that provides the movement vector of a spot
+	 * @return the relative movement vector of the spot
+	 */
+	public static double[] relativeMovement( final Spot spot, final int n, final Model model,
+			final Function< Spot, double[] > movementProvider )
```
[the remark this answers is not in the document]
Comment (stefanhahmann) [reply src/main/java/org/mastodon/mamut/feature/spot/SpotFeatureUtils.java:88]: I agree. I introduce a new method `neighborsAverageMovement`

File: src/test/java/org/mastodon/mamut/feature/branch/movement/relative/BranchRelativeMovementFeatureTest.java [hunk 9]
```diff
@@ -0,0 +1,111 @@
...
+		double actualZ = getProjection( branchRelativeMovementFeature, specZ ).value( graph3.branchSpotA );
+		double actualNorm = getProjection( branchRelativeMovementFeature, specNorm ).value( graph3.branchSpotA );
+
+		double[] expectedDirection = new double[] { 0d, 1d, 0d };
+		double expectedNorm = 2d;
+
+		assertEquals( expectedDirection[ 0 ], actualX, 0d );
+		assertEquals( expectedDirection[ 1 ], actualY, 0d );
+		assertEquals( expectedDirection[ 2 ], actualZ, 0d );
+		assertEquals( expectedNorm, actualNorm, 0d );
```
[the remark this answers is not in the document]
Comment (stefanhahmann) [reply src/test/java/org/mastodon/mamut/feature/branch/movement/relative/BranchRelativeMovementFeatureTest.java:73]: I decided to keep the simple test and added a 2nd test in BranchSpotFeatureUtilsTest that fulfills your request.

**Review by maarzt — APPROVED** [review]
The code is ok. I did a test run to and it works, as far as I can tell. However I would like to suggest some renaming:

I would not get what "Normalized relative movement norm to 5 nearest neighbors" means. Maybe "average speed relative to the 5 nearest neighbors".

There's the same problem for "Relative movement norm to 5 nearest neighbors" could we name it "movement distance relative to the 5 nearest neighbors".

**Comment by stefanhahmann** [comment]
> The code is ok. I did a test run to and it works, as far as I can tell. However I would like to suggest some renaming:
> 
> I would not get what "Normalized relative movement norm to 5 nearest neighbors" means. Maybe "average speed relative to the 5 nearest neighbors".
> 
> There's the same problem for "Relative movement norm to 5 nearest neighbors" could we name it "movement distance relative to the 5 nearest neighbors".

Thanks for the valuable comments regarding the naming. I implemented them.

**Pull request merged**